File: src/ctxdiff/capture/recorder.py
"""The fail-open recording path. Turns one (request, response) pair into a
stored call: adapter → token counts → hashes → labels → redaction → store.
Its single public method never raises — a debugging tool must not be able to
crash the program it is debugging."""
from __future__ import annotations

import copy
import logging
import threading
from dataclasses import dataclass
from typing import Callable

from ctxdiff.capture.base import Adapter
from ctxdiff.models import Block, CallBlock, basic_label, content_hash
from ctxdiff.store.ctrace import CTrace
from ctxdiff.tokenize.counter import count_tokens

_log = logging.getLogger("ctxdiff")


@dataclass(frozen=True)
class PersistJob:
    """One fully-built, fully-SNAPSHOTTED call ready to be written to the store.

    This is the hand-off between the two halves of recording (see `Recorder`):
    `build()` produces it on the CALLING thread — so every field is derived from
    the request/response AT CALL-COMPLETION TIME and owns its own immutable data
    (frozen dataclass, plain dicts, CallBlocks whose text is already copied out),
    with no lingering reference to the host's mutable `kwargs` — and `persist()`
    consumes it on the single WRITER thread. Because the snapshot is taken before
    the host can mutate its `messages`/`contents` list for the next turn (the
    normal agent-loop pattern of appending to one list and calling again), a
    deferred write can never record a later turn's contents against this seq."""
    seq: int
    params: dict
    usage: dict | None
    latency_ms: int | None
    error: str | None
    call_blocks: list[CallBlock]
    agent: str | None
    step: str | None
    provider: str | None


class Recorder:
    """Records calls into a CTrace using a provider adapter. Holds an optional
    redaction hook applied to every block just before it is stored."""

    def __init__(self, ctrace: CTrace, adapter: Adapter,
                 redact: Callable[[Block], Block] | None):
        """Wire the three collaborators `record()` needs: the store to write
        to, the provider-specific adapter that knows how to pull blocks/
        params/usage out of raw request/response objects, and an optional
        redaction hook. How: just stores the references; no I/O happens
        until `record()` is called."""
        self._ct = ctrace
        self._adapter = adapter
        self._redact = redact
        # One-time-warning state for persist failures (mirrors _Writer._degrade):
        # a persistently-broken store must warn ONCE for the run, not once per
        # call. Guarded by a lock because persist() runs on the writer thread.
        self._persist_warn_lock = threading.Lock()
        self._persist_warned = False

    def build(self, seq: int, kwargs: dict, response: object | None,
              latency_ms: int | None, error: str | None,
              tagged: list[tuple[str, str]],
              agent: str | None = None, step: str | None = None,
              provider: str | None = None, quiet: bool = False) -> PersistJob | None:
        """Phase 1 of recording — runs on the CALLING thread. What: turns a
        (request, response) pair into a fully self-contained `PersistJob`
        (blocks + params + usage + attribution), doing all the CPU work —
        adapter extraction, token counting, hashing, labelling, redaction — and,
        crucially, SNAPSHOTTING the request contents out of `kwargs` into owned,
        immutable data right here, before returning to the host. That snapshot
        is why recording can then be deferred to another thread without risk:
        the host may reuse/mutate its `messages` list for the next turn the
        instant this returns, but the block text was already copied out.

        How: mirrors the old inline pipeline exactly (so a sequentially-driven
        call produces a byte-identical stored row), but STOPS before touching
        the store — persistence is `persist()`'s job. Fail-open: any failure is
        logged once (unless `quiet`) and swallowed, returning None so the caller
        simply skips this call instead of the host ever seeing an error.
        `quiet` (set only by a stream proxy's best-effort `__del__` finalize —
        see trace.py) suppresses the warning at GC/interpreter-shutdown time,
        when logging is pure noise at best and can itself misbehave."""
        try:
            raw = self._adapter.extract_blocks(kwargs)
            # DEEP-COPY the params here, on the calling thread, before returning.
            # `extract_params` is a shallow dict comprehension whose values still
            # alias the host's kwargs objects; params is only json.dumps'd later,
            # on the writer thread. A host that passes a mutable param (metadata,
            # extra_body, stop, logit_bias, response_format) and mutates it
            # before that deferred write would otherwise corrupt this stored row.
            # deepcopy (not a json round-trip) so a non-JSON-serializable value
            # can't raise here and defeat the snapshot — and if deepcopy itself
            # ever raised, the surrounding fail-open guard still swallows it.
            params = copy.deepcopy(self._adapter.extract_params(kwargs))
            usage = self._adapter.extract_usage(response) if response is not None else None
            provider = self._adapter.provider

            call_blocks: list[CallBlock] = []
            for position, rb in enumerate(raw):
                # Count tokens for this text under the provider, then build the
                # content-addressed Block.
                token_count, token_method = count_tokens(rb.text, provider)
                block = Block(
                    content_hash=content_hash(rb.role, rb.kind, rb.text),
                    role=rb.role, kind=rb.kind, text=rb.text,
                    token_count=token_count, token_method=token_method,
                )
                # Redact after hashing/counting but before storage. The hash is
                # kept from the original text so identity/dedup is stable even if
                # redaction is nondeterministic; only the stored text changes.
                if self._redact is not None:
                    block = self._safe_redact(block)
                label, label_source = basic_label(rb.role, rb.kind, rb.text, tagged)
                call_blocks.append(CallBlock(
                    block=block, position=position,
                    label=label, label_source=label_source))

            return PersistJob(seq=seq, params=params, usage=usage,
                              latency_ms=latency_ms, error=error,
                              call_blocks=call_blocks,
                              agent=agent, step=step, provider=provider)
        except Exception:  # noqa: BLE001 — fail-open is the whole point
            if not quiet:
                _log.warning("ctxdiff: failed to build call seq=%s (tracing skipped)",
                             seq, exc_info=True)
            return None

    def persist(self, job: PersistJob, quiet: bool = False) -> None:
        """Phase 2 of recording — runs on the single WRITER thread. What: writes
        one already-built `PersistJob` to the store in a single transaction.
        How: a thin pass-through to `CTrace.record_call`; because every write
        for a run funnels through one writer thread, this is the ONLY place the
        connection is used for writing, so SQLite's thread-affinity holds even
        though the connection was opened on a different thread. Fail-open: a
        failed write is swallowed — a broken store must never take down the
        writer loop or, by extension, the host — and warned AT MOST ONCE for the
        run (see `_warn_persist_once`), so a persistently-broken store (disk
        full, read-only mount) can't flood the host's logs one line per call."""
        try:
            self._ct.record_call(seq=job.seq, params=job.params, usage=job.usage,
                                 latency_ms=job.latency_ms, error=job.error,
                                 call_blocks=job.call_blocks,
                                 agent=job.agent, step=job.step, provider=job.provider)
        except Exception:  # noqa: BLE001 — fail-open is the whole point
            if not quiet:
                self._warn_persist_once(job.seq)

    def _warn_persist_once(self, seq: int) -> None:
        """Emit the persist-failure warning AT MOST ONCE for this recorder, then
        stay silent — mirrors `_Writer._degrade`'s one-time mechanism. Why: a
        store that fails every write (disk full, read-only mount) would, with a
        per-call log, spam one warning line per recorded call; a lock-guarded
        `_persist_warned` flag makes only the FIRST failure log (with traceback)
        and every later one a no-op. Never host-facing regardless — this only
        ever reaches the `ctxdiff` logger."""
        with self._persist_warn_lock:
            if self._persist_warned:
                return
            self._persist_warned = True
        _log.warning("ctxdiff: failed to persist call seq=%s; further persist "
                     "failures in this run will be silent (tracing degraded)",
                     seq, exc_info=True)

    def record(self, seq: int, kwargs: dict, response: object | None,
               latency_ms: int | None, error: str | None,
               tagged: list[tuple[str, str]],
               agent: str | None = None, step: str | None = None,
               provider: str | None = None, quiet: bool = False) -> None:
        """Build AND persist one call inline, on the calling thread — the
        combined, synchronous path (`build()` then `persist()`). Retained for
        callers/tests that drive the whole pipeline in one step and don't route
        through the writer thread; the live capture path in trace.py instead
        calls `build()` on the host thread and hands `persist()` to the writer.
        Fail-open throughout via the two halves' own guards."""
        job = self.build(seq, kwargs, response, latency_ms, error, tagged,
                         agent, step, provider, quiet)
        if job is not None:
            self.persist(job, quiet)

    def _safe_redact(self, block: Block) -> Block:
        """Apply the redaction hook, but never let a throwing redactor break
        recording: on error, replace the text with a sentinel so nothing
        sensitive leaks and the run continues."""
        try:
            return self._redact(block)
        except Exception:  # noqa: BLE001
            _log.warning("ctxdiff: redact hook raised; using sentinel", exc_info=True)
            return Block(block.content_hash, block.role, block.kind,
                         "[redaction-error]", block.token_count, block.token_method)

File: js/src/capture/recorder.ts
/**
 * The fail-open recording path. Turns one (request, response) pair into a
 * stored call: adapter -> token counts -> hashes -> labels -> redaction ->
 * store. Its single public method never throws — a debugging tool must not be
 * able to crash the program it is debugging. Mirrors Python `Recorder`.
 */
import type { Adapter } from "./base.js";
import type { Block, CallBlock, RawBlock } from "../models.js";
import { basicLabel, contentHash } from "../models.js";
import { countTokens } from "../tokenize.js";
import type { Awaitable, RecordCallArgs, Store } from "../store/base.js";

export type RedactHook = (block: Block) => Block;

/**
 * Turn one adapter-extracted `RawBlock` into the stored, content-addressed
 * `Block`: give it its identity and its token numbers.
 *
 * Two overrides, both defaulting to the original behavior (see `RawBlock`):
 *
 *   - IDENTITY is `contentHash(role, kind, hashInput ?? text)`. For an ordinary
 *     block that is the text itself — unchanged, so every hash this SDK has ever
 *     written stays valid. For an image block it is a digest of the image BYTES,
 *     so two copies of the same picture are one block and a 1024×768 red square
 *     and a 1024×768 blue one are not.
 *   - TOKENS come from the tokenizer over `text`, unless the adapter already
 *     computed them — which it does only for images, where the truthful cost is
 *     the provider's vision formula over the pixel dimensions and NOT the
 *     tokenization of the `[image …]` descriptor standing in for them.
 *
 * Factored out of `Recorder.build` so there is exactly ONE definition of what a
 * stored block is: the golden harness (`test/helpers/golden.ts`) calls this same
 * function to materialize its fixtures, which is what lets the committed goldens
 * be evidence about the real capture path rather than about a parallel
 * reimplementation of it. Mirrors Python `build_block`.
 */
export function buildBlock(rb: RawBlock, provider: string): Block {
  const [tokenCount, tokenMethod] =
    rb.tokenCount !== undefined
      ? ([rb.tokenCount, rb.tokenMethod ?? "estimate"] as [number, string])
      : countTokens(rb.text, provider);
  const hashInput = rb.hashInput !== undefined ? rb.hashInput : rb.text;
  return {
    contentHash: contentHash(rb.role, rb.kind, hashInput),
    role: rb.role,
    kind: rb.kind,
    text: rb.text,
    tokenCount,
    tokenMethod,
  };
}

/**
 * One call, fully prepared and detached from the host's objects, ready to be
 * written. Produced by `build()` on the HOST's tick — blocks extracted, tokens
 * counted, hashes taken, labels resolved, redaction applied — and consumed by
 * `persist()`, which does nothing but hand it to the store.
 *
 * The split exists for the networked backends: a write there is a round trip, so
 * it happens off the host's call path (see `trace.ts`'s writer), and by then the
 * host may already have mutated the `messages` array the call was made with.
 * Snapshotting into a job at call time is what makes the deferred write record
 * what was actually SENT. It is exactly Python's `build`/`persist` pair.
 */
export type PersistJob = RecordCallArgs;

/** Whether `x` is promise-like — the one branch that lets this class serve both
 * a synchronous `CTrace` and a promise-based network store without either
 * paying for the other. */
function isThenable(x: unknown): x is Promise<unknown> {
  return (
    x != null &&
    (typeof x === "object" || typeof x === "function") &&
    typeof (x as { then?: unknown }).then === "function"
  );
}

export class Recorder {
  private store: Store;
  private adapter: Adapter;
  private redact: RedactHook | null;
  // One-time latch for the record-failure warning. A store that fails one write
  // usually fails every write (a full disk, a revoked file handle, a stuck
  // lock), and this path runs once per LLM call — unlatched it floods the host's
  // logs with an identical stack per turn. Mirrors Python's `_persist_warned`.
  private recordWarned = false;

  constructor(store: Store, adapter: Adapter, redact: RedactHook | null) {
    this.store = store;
    this.adapter = adapter;
    this.redact = redact;
  }

  /**
   * Build and store one call from its request kwargs and response. Every step
   * runs inside a catch-all: any failure is logged once and swallowed, leaving
   * the host's own call path untouched (fail-open). `tagged` is a list of
   * [label, needle] pairs used to override labels. `agent`/`step`/`provider`
   * are the v2 attribution fields threaded through to the store. `quiet`
   * suppresses the warning log (used only by a stream proxy's best-effort
   * finalize). Mirrors Python `Recorder.record`.
   *
   * This is the SYNCHRONOUS path, used with the local `.ctrace` store where the
   * write is a same-tick call: build the job, persist it, done — no promise is
   * created and nothing about the existing local behavior changes. A networked
   * store's writer calls `build()` and `persist()` separately instead.
   */
  record(args: {
    seq: number;
    kwargs: Record<string, unknown>;
    response: unknown;
    latencyMs: number | null;
    error: string | null;
    tagged: [string, string][];
    agent?: string | null;
    step?: string | null;
    provider?: string | null;
    quiet?: boolean;
  }): void {
    const job = this.build(args);
    if (job === null) return;
    const result = this.persist(job, args.quiet ?? false);
    // A synchronous store returns nothing; a store that unexpectedly returns a
    // promise here still cannot surface an unhandled rejection into the host.
    if (isThenable(result)) void result.then(undefined, () => undefined);
  }

  /**
   * Turn one (request, response) pair into a ready-to-write `PersistJob`, or
   * null if anything about the extraction fails (warned at most once, then
   * silent — fail-open). Pure and synchronous: it touches the adapter, the
   * tokenizer and the redaction hook, but never the store, so it is safe to run
   * on the host's own tick even when the store is a database on another
   * continent. Mirrors Python `Recorder.build`.
   */
  build(args: {
    seq: number;
    kwargs: Record<string, unknown>;
    response: unknown;
    latencyMs: number | null;
    error: string | null;
    tagged: [string, string][];
    agent?: string | null;
    step?: string | null;
    provider?: string | null;
    quiet?: boolean;
  }): PersistJob | null {
    const {
      seq,
      kwargs,
      response,
      latencyMs,
      error,
      tagged,
      agent = null,
      step = null,
      provider: providerArg = null,
      quiet = false,
    } = args;
    try {
      const raw = this.adapter.extractBlocks(kwargs);
      const params = this.adapter.extractParams(kwargs);
      const usage = response != null ? this.adapter.extractUsage(response) : null;
      // The caller's `provider` is the ATTRIBUTION label decided at wrap()
      // time — e.g. "gemini" for an OpenAI-SDK client pointed at Gemini's
      // OpenAI-compatible endpoint — and wins when supplied; the adapter's
      // own name is only the fallback for callers that don't attribute. Safe
      // because nothing mechanical below keys off this string: extraction
      // already ran above via the adapter, and `buildBlock` treats an
      // unrecognized provider as estimate-tokenized (which is MORE honest
      // for a non-OpenAI model than gpt-tokenizer counts marked exact).
      // Mirrors Python `Recorder.build` (dogfood finding 2026-07-27).
      const provider = providerArg ?? this.adapter.provider;

      const callBlocks: CallBlock[] = [];
      raw.forEach((rb, position) => {
        // Count tokens and take the content hash for this block — see
        // `buildBlock` for the two adapter-supplied overrides (an image hashes
        // its bytes and carries a pre-computed vision estimate; everything else
        // hashes and tokenizes its text).
        let block: Block = buildBlock(rb, provider);
        // Redact AFTER hashing/counting but BEFORE storage: the hash keeps the
        // original text so identity/dedup stays stable even if redaction is
        // nondeterministic; only the stored text changes.
        if (this.redact !== null) block = this.safeRedact(block);
        const [label, labelSource] = basicLabel(rb.role, rb.kind, rb.text, tagged);
        callBlocks.push({ block, position, label, labelSource });
      });

      return { seq, params, usage, latencyMs, error, callBlocks, agent, step, provider };
    } catch (err) {
      this.warnOnce(seq, err, quiet);
      return null;
    }
  }

  /**
   * Write one prepared job to the store. Returns whatever the store returns — a
   * promise for a networked backend (which the writer awaits, one job at a time,
   * so the single connection is never used concurrently) and nothing at all for
   * the local file. Fail-open in BOTH shapes: a synchronous throw and a rejected
   * promise are each caught here and warned about at most once, so a failing
   * store degrades capture and never reaches the host. Mirrors Python
   * `Recorder.persist`.
   */
  persist(job: PersistJob, quiet = false): Awaitable<void> {
    try {
      const result = this.store.recordCall(job);
      if (isThenable(result)) {
        return result.then(
          () => undefined,
          (err: unknown) => this.warnOnce(job.seq, err, quiet),
        );
      }
    } catch (err) {
      this.warnOnce(job.seq, err, quiet);
    }
  }

  /**
   * Report a capture failure AT MOST ONCE for the run, then stay silent: the
   * first failure carries the diagnosis (message + stack) and every subsequent
   * one is dropped rather than repeating the same stack once per turn. `quiet`
   * callers (a stream's best-effort finalize) skip logging entirely.
   */
  private warnOnce(seq: number, err: unknown, quiet: boolean): void {
    if (quiet || this.recordWarned) return;
    this.recordWarned = true;
    // eslint-disable-next-line no-console
    console.warn(
      `ctxdiff: failed to record call seq=${seq} (tracing skipped); ` +
        "further record failures in this run will be silent",
      err,
    );
  }

  /** Apply the redaction hook, but never let a throwing redactor break
   * recording: on error, replace the text with a sentinel so nothing sensitive
   * leaks and the run continues. Mirrors Python `_safe_redact`. */
  private safeRedact(block: Block): Block {
    if (this.redact === null) return block;
    try {
      return this.redact(block);
    } catch {
      return { ...block, text: "[redaction-error]" };
    }
  }
}
